webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java

```diff
@@ -52,7 +52,31 @@ public KademliaId(SMSPeer peer) {
         }
     }
 
-
+    /**
+     * Construct the NodeId from a string
+     *
+     * @param data The user generated key string
+     * @throws IllegalArgumentException If an id cannot be constructed from this string.
+     *                                  A NodeId is constructable from this string if the length
+     *                                  of the string is shorter than {@link #ID_LENGTH_BYTES}
+     *                                  or if the string is an hex representation of ad Id
+     *                                  (meaning it is long {@link #ID_LENGTH_BYTES} * 2 and the
+     *                                  only characters used are the numbers between 0 and 9
+     *                                  and the characters A through F)
+     * @author Marco Cognolato
+     */
+    public KademliaId(String data) {
+        // if it's in hex form, convert to byte array
+        if(data.matches("[0-9A-F]{" + ID_LENGTH_BYTES*2+"}")){
+            this.keyBytes = new byte[ID_LENGTH_BYTES];
+            for (int i = 0; i < keyBytes.length; i++) {
+                int index = i * 2;
+                int j = Integer.parseInt(data.substring(index, index + 2), 16);
+                keyBytes[i] = (byte) j;
+            }
```

Comment:
"[0-9A-F]" is the first part of a regex pattern. It's basically checking if every character is a character between 0-9 and A-F (or in another way to write it, if it's one of these characters "0123456789ABCDEF"), so I'm basically checking if evey character is a Hex representation of a number.
The other thing that pattern is checking (using that `"{" + ID_LENGTH_BYTES*2 +"}"` ) is check if the string is long ID_LENGTH_BYTES * 2, since 2 hex numbers are 1 byte, I need ID_LENGTH_BYTES * 2 to create a byte array ID_LENGTH_BYTES long.

The "16" is not quite "16 bits", it's meaning in the parseInt method is "base 16"

The "2" is used as follows:

the ID_LENGTH_BYTES * 2 has already been adressed ^

inside that for it's used in 2 ways.
2 hex caracters need to be transformed into a single number and then put into the keyBytes array, so:
I iterate over the keyBytes array length,
I take 2 characters at a time from the `data` String,
parse them from base 16 to int
and finally put them into the array at index `i`

So those "2" are:
`int index = i * 2` to convert from index of keyBytes to index of `data` where to take the string
and
`data.substring(index, index + 2)` where I take 2 characters of the `data` string at index `index`

Hope this was more clear

Reply:
Yes, very useful explanation. Thank you! 👍